### PR TITLE
Enhancement/lite db upgrade

### PR DESCRIPTION
**Pull request by manas-droid**
Thank you for your contribution.
Before submitting this PR, please make sure:
- [x] PR description and commit message should describe the changes done in this PR
- [x] Verify the PR is pointing to correct branch i.e. Release or Beta branch if the code fix is for specific release , else point it to master
- [x] Latest Code from master or specific release branch is merged to your branch
- [x] No unwanted\commented\junk code is included
- [x] No new warning upon build solution
- [ ] Code Summary\Comments are added to my code which explains what my code is doing
- [x] Existing unit test cases are passed
- [x] New Unit tests are added for your development
- [x] Sanity Tests are successfully executed for New and Existing Functionality
- [ ] Verify that changes are compatible with all relevant browsers and platforms.
- [x] After creating pull request there should not be any conflicts
- [x] Resolve codacy comments
- [ ] Builds and checks are passed before PR is sent for review
- [ ] Resolve code review comments
- [ ] Help Libray document is updated accordingly for Feature changes.

## Summary by CodeRabbit

- **New Features**
  - Improved data retrieval with new filtering capabilities.
  - Enhanced database interaction with the introduction of LiteDB SQL translation.

- **Enhancements**
  - Upgraded LiteDB package to enhance performance and stability.

- **Refactor**
  - Optimized data access by shifting to interface-based methods.
  - Streamlined data update processes in LiteDB operations.

- **Bug Fixes**
  - Fixed data filtering logic to ensure accurate data presentation.

- **Tests**
  - Updated unit tests to align with new LiteDB package version and functionalities.

### CODE REVIEW

**Comment by Maheshkale447** [comment]
> * @coderabbitai review

**Comment by manas-droid** [comment]
@Maheshkale447 I am yet to test if the enhancement works appropriately on Linux as we have found an existing issue with LiteDB v4 on Linux.

File: Ginger/GingerCoreNET/DataSource/LiteDBSQLTranslator.cs [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+        private void ValueExpressionForSetValue()
+        {
+            if (actDSTableElement.IsKeyValueTable)
+            {
+                GingerKeyValueSQLQuery.SetValueForGingerKeyValue(actDSTableElement , queryBuilder);
+            }
+            else if (actDSTableElement.Customized)
+            {
+                CustomizedSQLQuery.SetValueForCustomized(actDSTableElement, queryBuilder);
+            }
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/GingerCoreNET/DataSource/LiteDBSQLTranslator.cs:118]: @Maheshkale447 Do you want me make the queries as mentioned here?

File: Ginger/GingerCoreNET/DataSource/LiteDBSQLTranslator.cs [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+{
+    public class LiteDBSQLTranslator
+    {
+        private readonly ActDSTableElement actDSTableElement;
+        private StringBuilder queryBuilder;
+        public LiteDBSQLTranslator(ActDSTableElement actDSTableElement)
+        {
+            this.actDSTableElement = actDSTableElement;
+            queryBuilder = new ();
+        }
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/GingerCoreNET/DataSource/LiteDBSQLTranslator.cs:14]: This issue should not arise as there only 2 places where this class is used, 1. ActDSTableElement itself and another .xaml.cs class where ActDSTableElement is create in compile time

[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/GingerCoreNET/DataSource/LiteDBSQLTranslator.cs:14]: But if @Maheshkale447, you want I can add it anyways

File: Ginger/GingerCoreNETUnitTest/DataSource/DataSourceTest.cs [hunk 3]
```diff
@@ -247,7 +243,7 @@
         {
             //Arrange
             ActDSTableElement actDSTable = new ActDSTableElement();
-            string query = "db.MyCustomizedDataTable.find limit 1";
+            string query = "SELECT $ FROM MyCustomizedDataTable limit 1";
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/GingerCoreNETUnitTest/DataSource/DataSourceTest.cs:246]: According to the LiteDB documentation, this is the way to get all the columns from a table. Additionally this code exists in a Unit Test and all of our unit tests are passing.

File: Ginger/Ginger/RunSetPageLib/RunnerPage.xaml.cs [hunk 4]
```diff
@@ -433,7 +433,7 @@ private void GenerateBFReport(BusinessFlow bf)
                 LiteDbManager dbManager = new LiteDbManager(WorkSpace.Instance.Solution.SolutionOperations.ConvertSolutionRelativePath(WorkSpace.Instance.Solution.LoggerConfigurations.CalculatedLoggerFolder));
                 var result = dbManager.GetRunSetLiteData();
 
-                List<LiteDbRunSet> filterData = dbManager.FilterCollection(result, Query.All());
+                List<LiteDbRunSet> filterData = dbManager.FilterCollection<LiteDbRunSet>(LiteDbRunSet.IncludeAllReferences(result), Query.All());
 
                 LiteDbRunSet runSetLast = filterData.Last();
                 //runSetLast._id = new ObjectId();
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/Ginger/RunSetPageLib/RunnerPage.xaml.cs:439]: ok, will make the necessary changes

**Review by Maheshkale447 — DISMISSED** [review]

**Pull request merged**